# ##We need to importing the relevant libraries.
 

In [1]:
import pandas as pd
import  numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model  import LogisticRegression 

# #We load the Dataset

In [2]:
data=pd.read_csv("C:/Users/BENJAMIN/Downloads/dataset.csv")

In [3]:
data.info()  #Obtaining the information about the loaded dataset.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


In [4]:
data.head()  #Display the first 5 rows. 

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [5]:
columns=data.columns.values.tolist()  #Printing a list of all columns by their names.
print(columns)

['RowNumber', 'CustomerId', 'Surname', 'CreditScore', 'Geography', 'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember', 'EstimatedSalary', 'Exited']


## converting categorical columns to numeric columns In our dataset we have two categorical columns that is Geography and Gender.We need to convert them to numeric columns  



In [6]:
Geography=pd.get_dummies(data.Geography).iloc[:,1:]  #converting categorical columns to numerical values for easy training
Gender=pd.get_dummies(data .Gender).iloc[:,1:]
print(Geography)
print(Gender)

      Germany  Spain
0           0      0
1           0      1
2           0      0
3           0      0
4           0      1
...       ...    ...
9995        0      0
9996        0      0
9997        0      0
9998        1      0
9999        0      0

[10000 rows x 2 columns]
      Male
0        0
1        0
2        0
3        0
4        0
...    ...
9995     1
9996     1
9997     0
9998     1
9999     0

[10000 rows x 1 columns]


# After carefully observing the features we will remove the RowNumber, CustomerId, Surname columns and "Exited" column as they do not contribute to customer churning in any way


In [7]:
x=data.drop(["Exited",'RowNumber', 'CustomerId', 'Surname','Geography', 'Gender'],axis=1)
y=data["Exited"]
print(x)
print(y)

      CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0             619   42       2       0.00              1          1   
1             608   41       1   83807.86              1          0   
2             502   42       8  159660.80              3          1   
3             699   39       1       0.00              2          0   
4             850   43       2  125510.82              1          1   
...           ...  ...     ...        ...            ...        ...   
9995          771   39       5       0.00              2          1   
9996          516   35      10   57369.61              1          1   
9997          709   36       7       0.00              1          0   
9998          772   42       3   75075.31              2          1   
9999          792   28       4  130142.79              1          1   

      IsActiveMember  EstimatedSalary  
0                  1        101348.88  
1                  1        112542.58  
2                  0       

# #We need to Add the Geography and Gender data frames back to the dataset to create  the final dataset 


In [8]:
new =pd.concat([x,Geography,Gender],axis=1)
print(new)

      CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0             619   42       2       0.00              1          1   
1             608   41       1   83807.86              1          0   
2             502   42       8  159660.80              3          1   
3             699   39       1       0.00              2          0   
4             850   43       2  125510.82              1          1   
...           ...  ...     ...        ...            ...        ...   
9995          771   39       5       0.00              2          1   
9996          516   35      10   57369.61              1          1   
9997          709   36       7       0.00              1          0   
9998          772   42       3   75075.31              2          1   
9999          792   28       4  130142.79              1          1   

      IsActiveMember  EstimatedSalary  Germany  Spain  Male  
0                  1        101348.88        0      0     0  
1                  1   

# We can now split  into training set and test set.we will use 20% for test set and 80% for training set

In [9]:
x_train, x_test, y_train, y_test=train_test_split(x, y, test_size=0.2, random_state=0)
print(x_train, x_test, y_train, y_test)

      CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
7389          667   34       5       0.00              2          1   
9275          427   42       1   75681.52              1          1   
2995          535   29       2  112367.34              1          1   
5316          654   40       5  105683.63              1          1   
356           850   57       8  126776.30              2          1   
...           ...  ...     ...        ...            ...        ...   
9225          594   32       4  120074.97              2          1   
4859          794   22       4  114440.24              1          1   
3264          738   35       5  161274.05              2          1   
9845          590   38       9       0.00              2          1   
2732          623   48       1  108076.33              1          1   

      IsActiveMember  EstimatedSalary  
7389               0        163830.64  
9275               1         57098.00  
2995               0       

In [10]:
new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   CreditScore      10000 non-null  int64  
 1   Age              10000 non-null  int64  
 2   Tenure           10000 non-null  int64  
 3   Balance          10000 non-null  float64
 4   NumOfProducts    10000 non-null  int64  
 5   HasCrCard        10000 non-null  int64  
 6   IsActiveMember   10000 non-null  int64  
 7   EstimatedSalary  10000 non-null  float64
 8   Germany          10000 non-null  uint8  
 9   Spain            10000 non-null  uint8  
 10  Male             10000 non-null  uint8  
dtypes: float64(2), int64(6), uint8(3)
memory usage: 654.4 KB


## Machine learning algorithm training to indentify patterns or trends in the training dataset

In [11]:
from sklearn.linear_model  import LogisticRegression 
reg=LogisticRegression()
reg.fit(x_train,y_train)
predictions=reg.predict(x_test)

## Finding the accuracy of our logistic regression model.

In [12]:
from sklearn.metrics import confusion_matrix 
predictions=reg.predict(x_test)
confusion_matrix (y_test, predictions) 

array([[1553,   42],
       [ 380,   25]], dtype=int64)

## Evaluating logistic regression accuracy

In [282]:
from  sklearn.metrics import accuracy_score
accuracy_score(y_test, predictions)


0.789

# #In Conclusion,customer churn is crucial to the long-term financial stabilityof a company.we successfully created a machine learning algorithm that is able to predict customer churn with an accuracy of 78.9%